From 1109bd5a9ff4aa500c7ae825dfdb99c299143dfd Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Thu, 17 Mar 2022 23:17:13 +0100
Subject: [PATCH 19/19] winex11.drv: Improve fullscreen window detection
 heuristics.

For: Sleeping Dogs: Definitive Edition

CW-Bug-Id: #20283

 100.0% dlls/winex11.drv/
diff --git a/dlls/winex11.drv/window.c b/dlls/winex11.drv/window.c
index 36bbdc294e3..37737649125 100644
--- a/dlls/winex11.drv/window.c
+++ b/dlls/winex11.drv/window.c
@@ -1084,6 +1084,8 @@ static void update_net_wm_fullscreen_monitors( struct x11drv_win_data *data )
 void update_net_wm_states( struct x11drv_win_data *data )
 {
     UINT i, style, ex_style, new_state = 0;
+    RECT window_rect, client_rect;
+    BOOL fullscreen;
 
     if (!data->managed) return;
     if (data->whole_window == root_window)
@@ -1093,9 +1095,20 @@ void update_net_wm_states( struct x11drv_win_data *data )
     }
 
     style = NtUserGetWindowLongW( data->hwnd, GWL_STYLE );
+    ex_style = NtUserGetWindowLongW( data->hwnd, GWL_EXSTYLE );
+    if (!(fullscreen = NtUserIsWindowRectFullScreen( &data->whole_rect, get_win_monitor_dpi( data->hwnd ) )))
+    {
+        client_rect = data->client_rect;
+        OffsetRect( &client_rect, -client_rect.left, -client_rect.top );
+        window_rect = client_rect;
+        NtUserAdjustWindowRect( &window_rect, style, 0, ex_style, get_win_monitor_dpi( data->hwnd ) );
+        OffsetRect( &window_rect, data->window_rect.left - window_rect.left, data->window_rect.top - window_rect.top );
+        fullscreen = EqualRect( &window_rect, &data->window_rect ) && NtUserIsWindowRectFullScreen( &client_rect, get_win_monitor_dpi( data->hwnd ) );
+    }
+
     if (style & WS_MINIMIZE)
         new_state |= data->net_wm_state & ((1 << NET_WM_STATE_FULLSCREEN)|(1 << NET_WM_STATE_MAXIMIZED));
-    if (NtUserIsWindowRectFullScreen( &data->whole_rect, get_win_monitor_dpi( data->hwnd ) ))
+    if (fullscreen)
     {
         if ((style & WS_MAXIMIZE) && (style & WS_CAPTION) == WS_CAPTION)
             new_state |= (1 << NET_WM_STATE_MAXIMIZED);
-- 
2.43.0

